### PR TITLE
Supervised learning step

### PR DESCRIPTION
**Pull request by surya-money**
Added supervised step in `model.py` in `fit()` method before direction is chosen and made corresponding init changes to `Model` and `PercolatorModel` class and `brew()` method in `brew.py`. Also made necessary changes in `_find_hyperparameters()` method to properly utilize the supervised step added in `fit()`. Other minor changes in `model.py` include: changing estimator's refit parameter to true, fixing logger info in `_get_starting_labels()`, etc...

### CODE REVIEW

File: mokapot/model.py [hunk 1]
```diff
@@ -276,12 +280,33 @@ def fit(self, psms):
...
+        self.features = psms.features.columns.tolist()
+
+        # Supervised step
+        if self.supervised_step:
+            # Shuffle input order
+            size = psms.__len__()
+            my_shuffled_idx = np.random.permutation(np.arange(size))
+            my_original_idx = np.argsort(my_shuffled_idx)
+            shuffled_norm_feat = norm_feat[my_shuffled_idx, :]
+            shuffled_targets = psms.targets[my_shuffled_idx]
```
Comment (wfondrie) [inline mokapot/model.py:294]: We shouldn't always need to shuffle, and there is already shuffling code below. Remove this lines.

File: mokapot/model.py [hunk 2]
```diff
@@ -290,8 +315,8 @@ def fit(self, psms):
             norm_feat = norm_feat[shuffled_idx, :]
             start_labels = start_labels[shuffled_idx]
```
Comment (wfondrie) [inline mokapot/model.py:316]: Move this to above the added code.

File: mokapot/model.py [hunk 3]
```diff
@@ -290,8 +315,8 @@ def fit(self, psms):
...
 
-        # Prepare the model:
-        model = _find_hyperparameters(self, norm_feat, start_labels)
+        if not self.supervised_step:
+            model = _find_hyperparameters(self, norm_feat, start_labels)
```
Comment (wfondrie) [inline mokapot/model.py:319]: Remove this. We still need to find the optimal hyperparameters for the semi-supervised phase of model training.

File: mokapot/model.py [hunk 4]
```diff
@@ -421,14 +446,15 @@ def __init__(
         override=False,
         subset_max_train=None,
         n_jobs=-1,
+        supervised_step=False,
```
Comment (wfondrie) [inline mokapot/model.py:449]: We should remove this. I think we should keep `PercolatorModel` as the most accurate replication of Percolator as possible - that means excluding the `supervised_init`

If someone really wanted to, they could still enable it like this:
```python
model = PercolatorModel()
model.supervised_init = True
```

File: mokapot/model.py [hunk 5]
```diff
@@ -421,14 +446,15 @@ def __init__(
...
     ):
         """Initialize a PercolatorModel"""
         self.n_jobs = n_jobs
         svm_model = LinearSVC(dual=False)
         estimator = GridSearchCV(
             svm_model,
             param_grid=PERC_GRID,
-            refit=False,
+            refit=True,
```
Comment (wfondrie) [inline mokapot/model.py:457]: Why do we need to add this?

File: mokapot/model.py [hunk 6]
```diff
@@ -564,14 +591,15 @@ def _get_starting_labels(psms, model):
 
     elif model.is_trained:
```
Comment (wfondrie) [inline mokapot/model.py:592]: We still need to update the logic here. If our pre-trained model does not yield as many PSMs as the best feature, the best feature should be used and a warning message should be issued.

File: mokapot/model.py [hunk 7]
```diff
@@ -605,7 +633,7 @@ def _get_starting_labels(psms, model):
     return start_labels, feat_pass
 
 
-def _find_hyperparameters(model, features, labels):
+def _find_hyperparameters(model, features, labels, qvals=None):
```
Comment (wfondrie) [inline mokapot/model.py:636]: Add `qvals` documentation to the docstring.

File: mokapot/model.py [hunk 8]
```diff
@@ -628,7 +656,11 @@ def _find_hyperparameters(model, features, labels):
         cv_targ = (labels[labels.astype(bool)] + 1) / 2
 
         # Fit the model
-        model.estimator.fit(cv_samples, cv_targ)
+        if qvals is None:
+            model.estimator.fit(cv_samples, cv_targ)
+        else:
+            fit_params = {"sample_weight": (1 - qvals)}
+            model.estimator.fit(cv_samples, cv_targ, **fit_params)
```
Comment (wfondrie) [inline mokapot/model.py:663]: I'd simplify this to:
```python
weights = None if qvals is None else 1 - qvals
model.estimator.fit(cv_samples, cv_targ, sample_weight=weights)
```

**Review by wfondrie — CHANGES_REQUESTED** [review]
Overall this is looking pretty good, but I think there is some work to do. Please see below for my specific comments and let me know if you have any questions!

File: mokapot/model.py (hunk 5 above)
Comment (surya-money) [reply mokapot/model.py:457]: Right, we don't. This is redundant since we are manually refitting the best parameters in _find_hyperparameters()

File: mokapot/model.py (hunk 2 above)
Comment (surya-money) [reply mokapot/model.py:316]: I would need to move the entire code block under "Shuffle order" correct?

File: mokapot/brew.py [hunk 9]
```diff
@@ -14,7 +14,14 @@
...
 
 # Functions -------------------------------------------------------------------
-def brew(psms, model=None, test_fdr=0.01, folds=3, max_workers=1):
+def brew(
+    psms,
+    model=None,
+    test_fdr=0.01,
+    folds=3,
+    max_workers=1,
+    supervised_step=False,
```
[the remark this answers is not in the document]
Comment (surya-money) [reply mokapot/brew.py:23]: Good point, I only had that in for convenience when testing outside mokapot

File: mokapot/model.py [hunk 10]
```diff
@@ -17,6 +17,8 @@
 import logging
 import pickle
 import warnings
+import math
```
Comment (wfondrie) [inline mokapot/model.py:20]: Is this used anywhere?

File: mokapot/model.py [hunk 11]
```diff
@@ -86,6 +88,8 @@ class Model:
     shuffle : bool, optional
         Should the order of PSMs be randomized for training? For deterministic
         algorithms, this will have no effect.
+    supervised_init : bool, optional
+        Should the pre-training of the model include the fully supervised step?
```
Comment (wfondrie) [inline mokapot/model.py:92]: I'll update this to be a little more descriptive.

File: mokapot/model.py [hunk 12]
```diff
@@ -564,16 +583,32 @@ def _get_starting_labels(psms, model):
...
+                model.train_fdr,
+            )
+        else:
+            logging.warning(
+                "Learned model did not improve upon the pretrained "
+                "input model. Now re-scoring each collection of PSMs "
+                "using the best feature."
+            )
+            start_labels = start_labels_best_feat
+            feat_pass = feat_pass_best_feat
```
Comment (wfondrie) [inline mokapot/model.py:611]: This looks good, but it strikes me that maybe we should cache the results of `_find_best_feature` as part of the object. This is something I'll look into, so no action needed by you.

File: mokapot/brew.py [hunk 13]
```diff
@@ -14,7 +14,13 @@
...
 
 # Functions -------------------------------------------------------------------
-def brew(psms, model=None, test_fdr=0.01, folds=3, max_workers=1):
+def brew(
+    psms,
+    model=None,
+    test_fdr=0.01,
+    folds=3,
+    max_workers=1,
+):
```
Comment (wfondrie) [inline mokapot/brew.py:23]: Please revert this back to just a single line.

**Review by wfondrie — CHANGES_REQUESTED** [review]
This looks great to me @surya-money! There are just a couple minor changes I've noted to be fixed.

File: mokapot/model.py (hunk 10 above)
Comment (surya-money) [reply mokapot/model.py:20]: Nope! Had it in for testing, forgot to remove
